File: kotlin/demo/app/src/main/java/com/example/reduxforandroid/shoppingcart/domain/CartSlice.kt
package com.example.reduxforandroid.shoppingcart.domain

import android.util.Log
import com.example.reduxforandroid.Entity
import com.example.reduxforandroid.counter.CounterState
import com.example.reduxforandroid.counter.Decrement
import com.example.reduxforandroid.counter.Increment
import com.example.reduxforandroid.counter.LoadingComplete
import com.example.reduxforandroid.counter.StartLoading
import com.example.reduxforandroid.redux.Reducer
import com.example.reduxforandroid.todos.domain.AddTodo
import com.example.reduxforandroid.todos.domain.Todo
import com.example.reduxforandroid.todos.domain.ToggleTodo


data class Product(
    val id: String,
    val name: String,
    val price: String,
    val quantity: Int
)

data class ShoppingCart(
    val id: String,
    val items: List<String> //the list of product id
)



data class GetProductList(val keyword: String)

data class AddProductIntoCart(val productId: String,val shoppingCartId: String)

data class RemoveProductIntoCart(val productId: String,val shoppingCartId: String)

data class UpdateProductNumInCart(val productId: String,val num:Int,val shoppingCartId: String)

val productEntityReducer: Reducer<Entity<Product>> = { state, action ->
    when (action) {
        is GetProductList -> getProductListHandle(state,action)
        is UpdateProductNumInCart -> updateProductNumInCartHandle(state,action)
        else -> state
    }
}

val shoppingCartEntityReducer: Reducer<Entity<ShoppingCart>> = { state, action ->
    when (action) {
        is AddProductIntoCart -> addProductIntoCartHandle(state,action)
        is RemoveProductIntoCart -> removeProductIntoCartHandle(state,action)
        else -> state
    }
}

fun addProductIntoCartHandle(state: Entity<ShoppingCart>, action: AddProductIntoCart):Entity<ShoppingCart>{
    Log.d("CartSlice","action productId"+ action.productId + "shopping cartID"+ action.shoppingCartId )
    var tmp = state.byId.toMutableMap()
    var shoppingCart = tmp[action.shoppingCartId]
    if (shoppingCart == null)
        return state
    else{
        var itemList = shoppingCart.items.toMutableList()
        itemList.add(action.productId)
        val newShoppingCart = ShoppingCart(id = action.shoppingCartId, items = itemList)
        tmp[action.shoppingCartId] = newShoppingCart
        return state.copy(byId = tmp)
    }
}

fun removeProductIntoCartHandle(state: Entity<ShoppingCart>, action: RemoveProductIntoCart):Entity<ShoppingCart>{
    Log.d("CartSlice","action productId"+ action.productId + "action shoppingCartId" + action.shoppingCartId)
    var tmp = state.byId.toMutableMap()
    var shoppingCart = tmp[action.shoppingCartId]
    if (shoppingCart == null)
        return state
    else{
        var itemList = shoppingCart.items.toMutableList()
        itemList.remove(action.productId)
        val newShoppingCart = ShoppingCart(id = action.shoppingCartId, items = itemList)
        tmp[action.shoppingCartId] = newShoppingCart
        return state.copy(byId = tmp)
    }
}

fun getProductListHandle(state: Entity<Product>, action: GetProductList):Entity<Product>{
//    Log.d("CartSlice","action keyword"+ action.keyword)
    return state.copy(byId = mapOf(
        "product1" to Product(
            id = "product1",
            name = "user1",
            price = "20",
            quantity = 1
        ),
        "product2" to Product(
            id = "product2",
            name = "user2",
            price = "30",
            quantity = 2
        ),
    ),
        allIds = listOf("product1", "product2"));
}

fun replaceValueInMap(originalMap: Map<String, Product>, key: String, newValue: Product): Map<String, Product> {
    val newMap = originalMap.toMutableMap()
    newMap[key] = newValue
    return newMap
}

fun updateProductNumInCartHandle(state: Entity<Product>, action: UpdateProductNumInCart):Entity<Product>{
    Log.d("CartSlice","action productId"+ action.productId +" action num" + action.num)
    var tmp = state.byId
    var productEntity = tmp[action.productId]
    if (productEntity == null){
        return state;
    }
    else{
        val newProduct = Product(id = action.productId, name = productEntity.name, price = productEntity.price, quantity = action.num)
        val newMap = replaceValueInMap(tmp, action.productId, newProduct)
        return state.copy(byId = newMap);
    }

}

File: swift/demo/ReduxForIOS/ReduxForIOS/shoppingCart/CartSlice.swift
//
//  CartSlice.swift
//  ReduxForIOS
//
//  Created by JackLi on 2024/5/7.
//

import Foundation

struct Product : Equatable,CustomStringConvertible {
    var id: String
    var name: String
    var price: String
    var quantity: Int
    var description: String {
            return "id: \(id) name: \(name) price: \(price) quantity: \(quantity)"
        }
}

struct ShoppingCart : Equatable,CustomStringConvertible {
    var id: String
    var items: [String]
    var description: String {
            return "id: \(id)"
        }
}

struct GetProductList : Action{
    var keyword: String
}

struct AddProductIntoCart : Action{
    var productId: String
    var shoppingCartId: String
}

struct RemoveProductIntoCart : Action{
    var productId: String
    var shoppingCartId: String
}

struct UpdateProductNumInCart : Action{
    var productId: String
    var num: Int
    var shoppingCartId: String
}

func productEntityReducer(action: Action, state: Entity<Product>?) -> Entity<Product> {
    var state = state ?? Entity<Product>(byId: [String : Product](), allIds: [])
    switch action {
    case let action as GetProductList:
        state = getProductListHandle(state: state, action: action)
    case let action as UpdateProductNumInCart:
        state = updateProductNumInCartHandle(state: state, action: action)
    default:
        break
    }
    return state
}

func shoppingCartEntityReducer(action: Action, state: Entity<ShoppingCart>?) -> Entity<ShoppingCart> {
    var state = state ?? Entity<ShoppingCart>(byId: [String : ShoppingCart](), allIds: [])
    switch action {
    case let action as AddProductIntoCart:
        state = addProductIntoCartHandle(state: state, action: action)
    case let action as RemoveProductIntoCart:
        state = removeProductIntoCartHandle(state: state, action: action)
    default:
        break
    }
    return state
}



func addProductIntoCartHandle(state: Entity<ShoppingCart>, action: AddProductIntoCart) -> Entity<ShoppingCart>{
    print("CartSlice - action productId: \(action.productId), shopping cartID: \(action.shoppingCartId)")
    var tmp = state.byId
    let shoppingCart = tmp[action.shoppingCartId]
    if shoppingCart == nil {
        return state
    } else {
        var itemList = shoppingCart!.items
        itemList.append(action.productId)
        let newShoppingCart = ShoppingCart(id: action.shoppingCartId, items: itemList)
        tmp[action.shoppingCartId] = newShoppingCart
        return Entity<ShoppingCart>(byId: tmp, allIds: state.allIds)
    }
}

func removeProductIntoCartHandle(state: Entity<ShoppingCart>, action: RemoveProductIntoCart) -> Entity<ShoppingCart>{
    print("CartSlice - action productId: \(action.productId), action shoppingCartId: \(action.shoppingCartId)")
    var tmp = state.byId
    let shoppingCart = tmp[action.shoppingCartId]
    if shoppingCart == nil {
        return state
    } else {
        var itemList = shoppingCart!.items
        itemList.removeAll { $0 == action.productId }
        let newShoppingCart = ShoppingCart(id: action.shoppingCartId, items: itemList)
        tmp[action.shoppingCartId] = newShoppingCart
        return Entity<ShoppingCart>(byId: tmp, allIds: state.allIds)
    }
}

func getProductListHandle(state: Entity<Product>, action: GetProductList) -> Entity<Product> {
    return Entity(byId: [
        "product1": Product(id: "product1", name: "user1", price: "20", quantity: 1),
        "product2": Product(id: "product2", name: "user2", price: "30", quantity: 2)
    ], allIds: ["product1", "product2"])
}

func replaceValueInMap(originalMap: [String: Product], key: String, newValue: Product) -> [String: Product] {
    var newMap = originalMap
    newMap[key] = newValue
    return newMap
}

func updateProductNumInCartHandle(state: Entity<Product>, action: UpdateProductNumInCart) -> Entity<Product>{
    print("CartSlice - action productId: \(action.productId), action num: \(action.num)")
    let tmp = state.byId
    let productEntity = tmp[action.productId]
    if productEntity == nil {
        return state
    } else {
        let newProduct = Product(id: action.productId, name: productEntity!.name, price: productEntity!.price, quantity: action.num)
        let newMap = replaceValueInMap(originalMap: tmp, key: action.productId, newValue: newProduct)
        return Entity<Product>(byId: newMap, allIds: state.allIds)
    }
}


